### PR TITLE
feat: add new confirmation modal

### PR DESCRIPTION
**Pull request by arunjaindev**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] Does this PR require documentation updates?
- [ ] I've updated documentation as required by this PR.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/Shared/Helpers.tsx [hunk 1]
```diff
@@ -96,6 +98,14 @@ export const preventBodyScroll = (lock: boolean): void => {
     }
 }
 
+export const toggleOutsideFocus = ({ identifier, toggleFocus }: ToggleOutsideFocus) => {
```
Comment (AbhishekA1509) [inline src/Shared/Helpers.tsx:101]: Can we rename toggleFocus

File: src/Shared/Components/Backdrop/backdrop.scss [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.backdrop {
+  background: #000000bf;
```
Comment (AbhishekA1509) [inline src/Shared/Components/Backdrop/backdrop.scss:2]: Check if something is present in base for rgba

File: src/Shared/Components/Backdrop/Backdrop.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { ReactNode, useEffect } from 'react'
+import { motion } from 'framer-motion'
+import { useRegisterShortcut } from '@Common/Hooks'
+import { preventBodyScroll, toggleOutsideFocus } from '@Shared/Helpers'
+import './backdrop.scss'
+import { createPortal } from 'react-dom'
+import { ToggleFocusType } from '@Shared/types'
+import { DEVTRON_BASE_MAIN_ID } from '@Shared/constants'
+
+const Backdrop = ({ children, onEscape }: { children: ReactNode; onEscape: () => void }) => {
```
Comment (AbhishekA1509) [inline src/Shared/Components/Backdrop/Backdrop.tsx:10]: Please move type to props

File: src/Shared/Helpers.tsx [hunk 4]
```diff
@@ -96,6 +97,15 @@ export const preventBodyScroll = (lock: boolean): void => {
     }
 }
 
+export const preventOutsideFocus = ({ identifier, preventFocus }: ToggleOutsideFocus) => {
```
Comment (AbhishekA1509) [inline src/Shared/Helpers.tsx:100]: Please rename ToggleOutsideFocus

File: src/Shared/Components/ConfirmationModal/types.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+export enum ConfirmationModalVariantType {
+    info = 'info',
+    delete = 'delete',
+    warning = 'warning',
+    custom = 'custom',
+}
+
+interface CommonButtonProps extends Pick<ButtonProps, 'text'>, Partial<Pick<ButtonProps, 'startIcon' | 'endIcon'>> {
+    onClick: (e?: any) => void
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/types.tsx:12]: Can use SyntheticEvent

File: src/Shared/Components/ConfirmationModal/types.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          children: ReactNode
+      }
+    | {
+          customInputConfig?: never
+          children?: never
+      }
+
+export type ConfirmationModalProps = {
+    title: ReactNode
+    subtitle: ReactNode
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/types.tsx:50]: Can be optional

File: src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    subtitle,
+    Icon,
+    variant,
+    buttonConfig,
+    customInputConfig,
+    children,
+    showConfirmationModal,
+    handleClose,
+}: ConfirmationModalProps) => {
+    const { registerShortcut, unregisterShortcut } = useRegisterShortcut()
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx:22]: Please try to add some gap and maintain order between hook, states and variables

File: src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+
+    const handleCustomInputKeyDown = (e) => {
+        if (e.key === 'Enter') {
+            handleEnterKeyPress()
+        }
+    }
+
+    const handleCloseWrapper = useCallback(() => {
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx:43]: Are we sure about this check?

File: src/Shared/Components/Backdrop/backdrop.scss (hunk 2 above)
Comment (arunjaindev) [reply src/Shared/Components/Backdrop/backdrop.scss]: not present

File: src/Shared/Components/ConfirmationModal/types.tsx (hunk 5 above)
Comment (AbhishekA1509) [reply src/Shared/Components/ConfirmationModal/types.tsx:12]: Not yet done

File: src/Shared/Components/ConfirmationModal/types.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+          customInputConfig?: never
+          children: ReactNode
+      }
+    | {
+          customInputConfig?: never
+          children?: never
+      }
+
+export type ConfirmationModalProps = {
+    title: ReactNode
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Shared/Components/ConfirmationModal/types.tsx:49]: Not yet done

**Review by AbhishekA1509 — CHANGES_REQUESTED** [review]

File: src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        if (!primaryButtonConfig?.isLoading && !secondaryButtonConfig?.disabled) {
+            handleClose()
+        }
+    }, [primaryButtonConfig, secondaryButtonConfig])
+
+    useEffect(() => {
+        if (showConfirmationModal) {
+            // Timeout so that if modal is opened on enter press, it does not trigger onClick
+            setTimeout(() => {
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx:56]: Please add clearTimeout on unmount as well

File: src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+import { AnimatePresence, motion } from 'framer-motion'
+import { CustomInput, useRegisterShortcut, UseRegisterShortcutProvider } from '@Common/index'
+import { ComponentSizeType } from '@Shared/constants'
+import { ConfirmationModalProps } from './types'
+import { getPrimaryButtonStyleFromVariant, getConfirmationLabel, getIconFromVariant } from './utils'
+import { Button, ButtonStyleType, ButtonVariantType } from '../Button'
+import './confirmationModal.scss'
+import { Backdrop } from '../Backdrop'
+
+let timeoutId: number
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx:11]: We can move this to ref

File: src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            // Timeout so that if modal is opened on enter press, it does not trigger onClick
+            timeoutId = setTimeout(() => {
+                registerShortcut({ keys: ['Enter'], callback: handleTriggerPrimaryActionButton })
+            }, 100)
+        }
+
+        return () => {
+            if (showConfirmationModal) {
+                unregisterShortcut(['Enter'])
+                clearTimeout(timeoutId)
```
Comment (AbhishekA1509) [inline src/Shared/Components/ConfirmationModal/ConfirmationModal.tsx:66]: This should be outside this check ideally check should be if (timeoutId) { clearTimeout(timeoutId) }

**Review by AbhishekA1509 — DISMISSED** [review]

**Review by AbhishekA1509 — APPROVED** [review]

**Review by AbhishekA1509 — APPROVED** [review]
Please add description to PR template

**Review by AbhishekA1509 — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**